### PR TITLE
Added checks that selection is valid, and fix set_selection_range.

### LINKED ISSUE #11901: Panic when deleting text on bhtml url bar
Running from bhtml master with:
`./mach run -d -- -b -w --resolution=1024x720 --pref dom.mozbrowser.enabled --pref dom.forcetouch.enabled --pref dom.quit-on-escape.enabled=false http://localhost:6060`

1. Type text in the URL bar
2. Hover over suggestion
3. Hit delete until panic

```
Render took 64.79874900000004ms & will cause frame drop
Unhandled error occured
ReferenceError: navigator.mozSettings API is not available
Render took 64.90834999999998ms & will cause frame drop
Render took 38.891733000000386ms & will cause frame drop
ERROR:constellation::constellation: Panic: begin <= end (3 <= 1) when slicing `a`
ERROR:constellation::constellation: Backtrace:
frame #0  - 0x000000010fb3a780 - backtrace::backtrace::trace::hccde8df28b4db2a2
frame #1  - 0x000000010fb3a62a - backtrace::capture::Backtrace::new::h42f95930bb8c5ee8
frame #2  - 0x000000010e2b3355 - util::thread::spawn_named_with_send_on_panic::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h8e08bac0ce5835b1
frame #3  - 0x000000010e2b3950 - extern $u22$rust.call$u22$$u20$fn$LP$$u5b$closure$SP$DefId$u20$$u7b$$u20$krate.$u20$47$C$$u20$node.$u20$DefIndex$LP$2380$RP$$u20$$u7d$$u20$0.ipc_channel..ipc..IpcSender$LT$$LP$core..option..Option$LT$msg..constellation_msg..PipelineId$GT$$C$$u20$std..string..String$C$$u20$std..string..String$RP$$GT$$C$$u20$1.core..option..Option$LT$msg..constellation_msg..PipelineId$GT$$u5d$$C$$u20$$LP$$RF$core..any..Any$C$$RP$$RP$::once_shim.210704::h3190046b5fea3554
frame #4  - 0x000000010e2b38fc - _<F as alloc..boxed..FnBox<A>>::call_box::hebb3b95724d7c6a0
frame #5  - 0x000000010f9ff63f - util::panicking::initiate_panic_hook::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::hddf831bcfa835e79
frame #6  - 0x000000010f9ff243 - _<std..thread..local..LocalKey<T>>::with::h49e68c592f601e39
frame #7  - 0x000000010f9ff096 - util::panicking::initiate_panic_hook::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h0fcde37d4d058867
frame #8  - 0x000000010feac714 - std::panicking::rust_panic_with_hook::h983af77c1a2e581b
frame #9  - 0x000000010fec57d6 - std::panicking::begin_panic::he426e15a3766089a
frame #10 - 0x000000010feadd58 - std::panicking::begin_panic_fmt::hdddb415186c241e7
frame #11 - 0x000000010fec542f - rust_begin_unwind
frame #12 - 0x000000010feec2b0 - core::panicking::panic_fmt::hf4e16cb7f0d41a25
frame #13 - 0x000000010fef428a - core::str::slice_error_fail::ha8f4daeb58b867c4
frame #14 - 0x000000010ce2b52c - core::str::traits::_<impl core..ops..Index<core..ops..RangeFrom<usize>> for str>::index::hbdc52817d320a9bf
frame #15 - 0x000000010dfc0449 - _<script..textinput..TextInput<T>>::replace_selection::ha5616f5185516242
frame #16 - 0x000000010dfc45f2 - _<script..textinput..TextInput<T>>::delete_char::h6a3dc6d05bc6fd31
frame #17 - 0x000000010dfbeb4e - _<script..textinput..TextInput<T>>::handle_keydown_aux::h53780aad87c6bdca
frame #18 - 0x000000010dfbe75e - _<script..textinput..TextInput<T>>::handle_keydown::h04b717b2a3e0a2d5
frame #19 - 0x000000010dfbe284 - _<script..dom..htmlinputelement..HTMLInputElement as script..dom..virtualmethods..VirtualMethods>::handle_event::h86107a9c7fec09db
frame #20 - 0x000000010df16508 - script::dom::eventdispatcher::dispatch_event::h2af7a9aca7ed0603
frame #21 - 0x000000010ce9c413 - script::dom::eventtarget::EventTarget::dispatch_event::h81415f3730b3c0d9
frame #22 - 0x000000010de513bc - script::dom::event::Event::fire::hc8f0f175fa50a18c
frame #23 - 0x000000010de58d41 - script::dom::document::Document::dispatch_key_event::hf0b3c82ee1e5d75d
frame #24 - 0x000000010e332e3d - script::script_thread::ScriptThread::handle_event::hb1054401ed8869b2
frame #25 - 0x000000010e363127 - script::script_thread::ScriptThread::handle_msg_from_constellation::hb125e29adccf30ce
frame #26 - 0x000000010e361bde - script::script_thread::ScriptThread::handle_msgs::_$u7b$$u7b$closure$u7d$$u7d$::hdfba591677ee0854
frame #27 - 0x000000010e3617c5 - script::script_thread::ScriptThread::profile_event::h180ff2036b8c2783
frame #28 - 0x000000010e32f6dd - script::script_thread::ScriptThread::handle_msgs::h45a47f1f1786f911
frame #29 - 0x000000010e2c80bc - script::script_thread::ScriptThread::start::hf0f02e16b9da3912
frame #30 - 0x000000010e2c806b - _<script..script_thread..ScriptThread as script_traits..ScriptThreadFactory>::create::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h60e2514d7c15aeaf
frame #31 - 0x000000010e2c7dc1 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::h90164fdc02bc9ddf
frame #32 - 0x000000010e2b4076 - _<script..script_thread..ScriptThread as script_traits..ScriptThreadFactory>::create::_$u7b$$u7b$closure$u7d$$u7d$::hc2f6bfd107201055
frame #33 - 0x000000010e2b270e - util::thread::spawn_named_with_send_on_panic::_$u7b$$u7b$closure$u7d$$u7d$::h44d46100d4f5f190
frame #34 - 0x000000010e2b24c7 - _<std..panic..AssertUnwindSafe<F> as core..ops..FnOnce<()>>::call_once::h4974dfb1034d8cd6
frame #35 - 0x000000010e2b241b - std::panicking::try::_$u7b$$u7b$closure$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h4326e3cb5bf9d187
frame #36 - 0x000000010e2b2cd4 - std::panicking::try::call::h78061d8f9ce27ca7
frame #37 - 0x000000010fec84bb - __rust_try
frame #38 - 0x000000010fec8455 - __rust_maybe_catch_panic
frame #39 - 0x000000010e2b236d - std::panicking::try::_$u7b$$u7b$closure$u7d$$u7d$::he258e52e28311769
frame #40 - 0x000000010e2b22bf - _<std..thread..local..LocalKey<T>>::with::h6a07862be6a45706
frame #41 - 0x000000010e2b20ef - std::panicking::try::h90b38a4a00a80b43
frame #42 - 0x000000010e2b1fce - std::panic::catch_unwind::h066dae249590ffb5
frame #43 - 0x000000010e2b1e2c - std::thread::Builder::spawn::_$u7b$$u7b$closure$u7d$$u7d$::hb658a84aa2bff2ef
frame #44 - 0x000000010e2b2ed7 - _<F as alloc..boxed..FnBox<A>>::call_box::h2ec158fb92e41aca
frame #45 - 0x000000010fec3f38 - std::sys::thread::Thread::new::thread_start::h9c883b6d445ece46
frame #46 - 0x00007fff9eeb699c - _pthread_body
frame #47 - 0x00007fff9eeb6919 - _pthread_start

ERROR:constellation::constellation: Panic: called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: Other, error: StringError("MPSC send failed") }) }
```

### PR DESCRIPTION
**Pull request by asajeffrey**
Fixed range checking for text inputs. Included some `debug_assert!` statements to check that the selection range is always valid.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11901.
- [X] These changes do not require tests because the code includes debug_asserts to ensure invariants are maintained.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11906)

### CODE REVIEW

**Comment by ConnorGBrewster** [comment]
@bors-servo retry
 * #11703

**Comment by jdm** [comment]
No test?

**Comment by ConnorGBrewster** [comment]
@jdm yeah might be good to add a test. Would a unit test work?

**Comment by jdm** [comment]
If a unit test can reproduce the problem, I'm fine with it.

**Comment by asajeffrey** [comment]
@jdm: I added a bunch of `debug_assert!`s, does that count?

**Comment by Ms2ger** [comment]
We run wpt in release, so not sure how useful those are if you don't write a unit test.

**Comment by asajeffrey** [comment]
@jdm @Ms2ger okay, I'll see if I can work out how to get a wpt test to trigger this (it's slightly tricky, because it's easy to trigger by user input, but I'm not sure we can trigger it from JS).

**Comment by asajeffrey** [comment]
Actually, there's a pretty direct code path from https://html.spec.whatwg.org/multipage/forms.html#dom-textarea/input-setselectionrange to `set_selection_range`. I'm a bit surprised there's not a test for invalid values already.

**Comment by asajeffrey** [comment]
Some IRC chat: http://logs.glob.uno/?c=mozilla%23servo&s=29+Jun+2016&e=29+Jun+2016#c465777

TL;DR: unsigned long parameters to webIDL functions are truncated to the range `0..2^32-1`

**Comment by asajeffrey** [comment]
Squashed.

@bors-servo r=ConnorGBrewster

**Comment by ConnorGBrewster** [comment]
@bors-servo retry
 * #11703
